lib/core/figbird.js

```diff
@@ -0,0 +1,604 @@
+      this.#listeners.set(queryId, new Set())
+    }
+    this.#listeners.get(queryId).add(fn)
+    return () => {
+      const listeners = this.#listeners.get(queryId)
+      if (listeners) {
+        listeners.delete(fn)
+        if (listeners.size === 0) {
+          this.#listeners.delete(queryId)
+        }
+      }
+    }
+  }
+
+  #invokeListeners(queryId) {
+    const listeners = this.#listeners.get(queryId)
+    if (listeners) {
+      const state = this.getQueryState(queryId)
+      listeners.forEach(listener => listener(state))
+    }
+  }
+
+  #addGlobalListener(fn) {
+    this.#globalListeners.add(fn)
+    return () => {
+      this.#globalListeners.delete(fn)
+    }
+  }
+
+  #invokeGlobalListeners() {
+    const state = this.getState()
+    this.#globalListeners.forEach(listener => listener(state))
+  }
+
+  #listenerCount(queryId) {
+    return this.#listeners.get(queryId)?.size || 0
+  }
+
+  subscribe(queryId, fn) {
+    const q = this.#getQuery(queryId)
+    if (
+      q.pending ||
+      (q.state.status === 'success' && q.config.fetchPolicy === 'swr' && !q.state.isFetching) ||
+      (q.state.status === 'error' && !q.state.isFetching)
+    ) {
+      this.#queue(queryId)
+    }
+
+    const removeListener = this.#addListener(queryId, fn)
+
+    this.#subscribeToRealtime(queryId)
+
+    const shouldVacuumByDefault = q.config.fetchPolicy === 'network-only'
+    return ({ vacuum = shouldVacuumByDefault } = {}) => {
+      removeListener()
+      if (vacuum && this.#listenerCount(queryId) === 0) {
+        this.#vacuum({ queryId })
+      }
+    }
+  }
+
+  subscribeToStateChanges(fn) {
+    return this.#addGlobalListener(fn)
+  }
+
+  refetch(queryId) {
+    const q = this.#getQuery(queryId)
+    if (!q.state.isFetching) {
+      this.#queue(queryId)
+    }
+  }
+
+  async #queue(queryId) {
+    this.#fetching({ queryId })
+    try {
+      const result = await this.#fetch(queryId)
+      this.#fetched({ queryId, result })
+    } catch (err) {
+      this.#fetchFailed({ queryId, error: err })
+    }
+  }
+
+  #fetch(queryId) {
+    const query = this.#getQuery(queryId)
+    const { serviceName, method, resourceId, params } = query.desc
+    const { allPages } = query.config
+
+    if (method === 'get') {
+      return this.#adapter.get(serviceName, resourceId, params)
+    } else if (allPages) {
+      return this.#adapter.findAll(serviceName, params)
+    } else {
+      return this.#adapter.find(serviceName, params)
+    }
+  }
+
+  mutate({ serviceName, method, args }) {
+    const updaters = {
+      create: item => this.#created({ serviceName, item }),
+      update: item => this.#updated({ serviceName, item }),
+      patch: item => this.#patched({ serviceName, item }),
+      remove: item => this.#removed({ serviceName, item }),
+    }
+
+    return this.#adapter.mutate(serviceName, method, args).then(item => {
+      updaters[method](item)
+    })
+  }
+
+  #subscribeToRealtime(queryId) {
```

Comment:
sometimes this returns an unsubscribe fn but if it's already subscribed it will return undefined. And it looks like where this is called it never uses the returned value so maybe missing an unsubscribe call somewhere?

If the idea is to subscribe once but allow multiple calls might need a reference counter?

Reply:
Yeah, this bit a bit unfinished - the idea right now is that we never unsubscribe from relatime (and never delete anything from the cache) - once you ran the query, it keeps consuming realtime events. How to unsub/evict - not currently defined in. This is not far off from what the app is already doing, so I don't see this to be a big issue.